core/lib/multivm/src/tracers/prestate_tracer/vm_virtual_blocks/mod.rs

```diff
@@ -0,0 +1,72 @@
+use zk_evm_1_3_3::tracing::{BeforeExecutionData, VmLocalStateData};
+use zksync_state::{StoragePtr, WriteStorage};
+use zksync_types::StorageKey;
+
+use super::{
+    get_account_data, process_modified_storage_keys, process_result, PrestateTracer, State,
+    StorageAccess,
+};
+use crate::{
+    interface::dyn_tracers::vm_1_3_3::DynTracer,
+    tracers::prestate_tracer::U256,
+    vm_virtual_blocks::{
+        BootloaderState, ExecutionEndTracer, ExecutionProcessing, HistoryMode, SimpleMemory,
+        ZkSyncVmState,
+    },
+};
+
+impl<S: WriteStorage, H: HistoryMode> DynTracer<S, SimpleMemory<H>> for PrestateTracer {
+    fn before_execution(
+        &mut self,
+        _state: VmLocalStateData<'_>,
+        _data: BeforeExecutionData,
+        _memory: &SimpleMemory<H>,
+        storage: StoragePtr<S>,
+    ) {
+        if self.config.diff_mode {
+            self.pre
+                .extend(process_modified_storage_keys(self.pre.clone(), &storage));
+        }
+    }
+}
+
+impl<H: HistoryMode> ExecutionEndTracer<H> for PrestateTracer {}
+
+impl<S: WriteStorage, H: HistoryMode> ExecutionProcessing<S, H> for PrestateTracer {
+    fn after_vm_execution(
+        &mut self,
+        state: &mut ZkSyncVmState<S, H>,
+        _bootloader_state: &BootloaderState,
+        _stop_reason: crate::interface::tracer::VmExecutionStopReason,
+    ) {
+        let modified_storage_keys = state.storage.storage.inner().get_modified_storage_keys();
+        if self.config.diff_mode {
+            self.post = modified_storage_keys
+                .clone()
+                .keys()
+                .copied()
+                .collect::<Vec<_>>()
+                .iter()
+                .map(|k| get_account_data(k, state, &modified_storage_keys))
+                .collect::<State>();
+        } else {
+            let read_keys = &state.storage.read_keys;
+            let map = read_keys.inner().clone();
+            let res = map
+                .keys()
+                .copied()
+                .collect::<Vec<_>>()
+                .iter()
+                .map(|k| get_account_data(k, state, &modified_storage_keys))
+                .collect::<State>();
+            self.post = res;
+        }
+        process_result(&self.result, self.pre.clone(), self.post.clone());
+    }
+}
+
+impl<S: zksync_state::WriteStorage, H: HistoryMode> StorageAccess for ZkSyncVmState<S, H> {
+    fn read_from_storage(&self, key: &StorageKey) -> U256 {
+        self.storage.storage.read_from_storage(key)
+    }
+}
```

Comment:
tbh, i'd rather expose `storage_ptr` to after_vm _execution

Reply:
Yeah I see the merit in directly passing `storage_ptr` to enhance the function 🙏 . However, implementing this across all after_execution instances could significantly broaden the current PR's scope. Maybe we can group the changes in a separate PR?